keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+In the above figures, the logic of `ImagePullManager.getRecord()` first attempts
+to retrieve a record of a successfully *pulled* image, and if it does not find any
+it will also attempt to find a record of an image currently being in a *pulling*
+state.
+
+##### Failure modes:
+
+We should always fail safe. If there's an error reading from the disk, the record
+MUST be assumed as non-existent, meaning the image pull needs credential re-verification.
+This might cause an additional request to pull an image. If it is successful,
+the original failing record should be overridden to reflect the new success.
+
+#### Cache Directory Structure
+
+A new subdirectory is introduced to the kubelet's directory: `image_manager`.
+The directory contains two further subdirectories: `pulling` and `pulled`. The former
+serves for kubelet to store the state about images that it intends to pull. The
+latter subdirectory should contain information about all the images that were
+successfully pulled by the kubelet.
+
+The figure below shows the Go structures that should appear in files in the
+subdirectories described above. These structures **are not part of the REST API**
+but belong to the `imagemanager.kubelet.config.k8s.io` configuration API group.
+They only ever appear as JSON structures in files outside the kubelet memory.
+
+`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
+directory. The filename of each such file is a SHA-256 hash of the image spec prepended
+by `sha256~`.
+
+`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
+directory. The filename of each such file is a sanitized image ID provided by the
+container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,
+thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.
+
+```go
+// ImagePullIntent is a record of the kubelet attempting to pull an image.
+type ImagePullIntent struct {
+    metav1.TypeMeta
+
+    // Image is the image spec from a Container's `image` field.
+    // The filename is a SHA-256 hash of this value. This is to avoid filename-unsafe
+    // characters like ':' and '/'.
+    Image string `json:"image"`
+}
+
+// ImagePullRecord is a record of an image that was pulled by the kubelet.
+//
+// If there are no records in the `kubernetesSecrets` field and both `nodeWideCredentials`
+// and `anonymous` are `false`, credentials must be re-checked the next time an
+// image represented by this record is being requested.
+type ImagePulledRecord struct {
+    metav1.TypeMeta
+
+    // ImageRef is a reference to the image represented by this file as received
+    // from the CRI.
+    // The filename is a filesystem-sanitized form of this value.
+    ImageRef string `json:"imageRef"`
+
+    // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes
```

Comment:
nest / partition the credential info under a map keyed by the `registry/repo/image` that was pulled (can drop the tag and the digest from the image spec)

```
// map of imagespec (without tag/digest) to credentials valid for that imagespec
CredentialInfo map[string]CredentialInfo
```

This ensures that an anonymous pod can't reference a private registry or repo that *happens* to have a tagged public image in it, and discover that exists and what content it identifies.